modules/core/arrow-generic/src/test/kotlin/arrow/generic/CoproductTest.kt

```diff
@@ -82,5 +82,37 @@ class CoproductTest : UnitSpec() {
                     { "Third" }
             ) shouldBe "Third"
         }
+
+        "cop replacement functions are equivalent" {
+            val firstValue = "String"
+            val secondValue = 100L
+            val thirdValue = none<String>()
+
+            firstValue.cop<String, Long, Option<String>>() shouldBe firstValue.first<String, Long, Option<String>>()
+            secondValue.cop<String, Long, Option<String>>() shouldBe secondValue.second<String, Long, Option<String>>()
+            thirdValue.cop<String, Long, Option<String>>() shouldBe thirdValue.third<String, Long, Option<String>>()
+        }
+
+        "types can be inferred with the cop replacement functions" {
+            fun typeInferenceTest(input: String?): Coproduct3<String, Int, Option<String>> {
+                return when {
+                    input == null -> none<String>().third()
+                    input.length > 100 -> input.length.second()
+                    else -> input.first()
+                }
+            }
+
+            typeInferenceTest(null) shouldBe Third<String, Int, Option<String>>(none())
+        }
+
+        "Coproduct data classes should be available" {
+            val firstValue = "String"
+            val secondValue = 100L
+            val thirdValue = none<String>()
+
+            First<String, Long, Option<String>>(firstValue) shouldBe firstValue.first<String, Long, Option<String>>()
```

Comment:
A similar one but with a function to test type inference by using the ADT constructors would be nice